### PR TITLE
Status tracking of remote nodes

### PR DESCRIPTION
**Pull request by kyessenov**
- Adds latest ack/push version tracking for each node
- Prevent repeated push of invalid configs and driving Envoy into live-lock

### CODE REVIEW

File: pkg/cache/status.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+		// there are two main causes of rejection by the proxy:
+		// - internal validity is violated:
+		//   the response fails to validate and repeated responses do not lead to progress
+		// - referential validity is violated (e.g. validate_clusters option):
+		//   retries may lead to progress but this requires Envoy to enforce referential validity (a non-default behavior)
+		//   it is important to stagger updates and not immediately reply in this case
+		return false
+	}
+
+	ni.responded[typ] = version
```
Comment (rshriram) [inline pkg/cache/status.go]: If this is the only place where this is being updated, do we even need a lock ? I guess what I am driving at is whether this code is in the fast path (though it’s funny to say fast path in control plane code)

File: pkg/cache/status.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+	// latest responded version
+	responded map[ResponseType]string
+}
+
+// StatusInfo is a status tracker for configuration version acknowledgement by
+// individual proxy nodes.
+type StatusInfo struct {
+	id      NodeID
+	compare VersionCompare
```
[the remark this answers is not in the document]
Comment (9len) [reply pkg/cache/status.go]: (just go be a golang pedant, the name should actually be `nodeID`)

File: pkg/cache/status.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+// VersionCompare compares version identifiers and returns of three values, +1 if
+// if first argument is greater than the second, 0 if equal, and -1 if less.
+type VersionCompare func(string, string) int
+
+// NodeInfo is node status.
+type NodeInfo struct {
+	id string
+	mu sync.RWMutex
```
Comment (bndw) [inline pkg/cache/status.go]: I believe the idiomatic way to use locks is to nest them

```go
type NodeInfo struct {
    sync.RWMutex
}

func (n *NodeInfo) Foo() {
    n.Lock()
    n.Unlock()
}
```

File: pkg/cache/status.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+		id:      id,
+		compare: compare,
+		nodes:   make(map[string]*NodeInfo),
+	}
+}
+
+// GetInfo retrieves node info or creates a record for the node.
+func (info *StatusInfo) GetInfo(node *api.Node) *NodeInfo {
+	info.mu.Lock()
+	defer info.mu.Unlock()
```
[the remark this answers is not in the document]
Comment (bndw) [reply pkg/cache/status.go]: If we want a read lock here it should be `RLock()` and `RUnlock()`

**Pull request closed without merging**